Add thumbnail generation integration test

goto-bus-stop opened the pull request:
Depends on https://github.com/transloadit/uppy/pull/968

To make this test easier to write, I added some events to the ThumbnailGenerator plugin. People had been asking for that so I guess it'll be useful for them too :P

It just adds some files using a bare `@uppy/file-input` and checks that thumbnails of the right size are being added to the DOM.

TODO:

 - [x] Add some non image files to check that they're not being touched
 - [x] Add some corrupt image files to test error handling

Fixes #946

arturi commented on packages/@uppy/thumbnail-generator/src/index.js:
```diff
@@ -186,9 +186,11 @@ module.exports = class ThumbnailGenerator extends Plugin {
         })
         .catch(err => {
           console.warn(err.stack || err.message)
+          this.uppy.emit('thumbnail:error', this.uppy.getFile(file.id), err)
```
So `thumbnail:ready` is when all are ready, and `thumbnail:generated` is for single? Maybe add `all-` to it then, `thumbnail:all-ready` or `thumbnail:all-generated`?

arturi replied:
Had the ready event in a local file, was about to submit a PR to address #946, but this adds errors and all too, cool 😎

jazkh commented:
I was about to ask the same thing if i could use file thumbnail to send it to the server along with the original file. But i couldn't figure it out how to get access to the thumbnail after successfull upload.

goto-bus-stop commented:
Should be ready to go (pending CI)

It also adds a (hacky) `-b` flag to the `test:acceptance:local` script, so you can test in different browsers without editing the wdio config. default is just FF

```bash
# test thumbnails in chrome and FF
npx wdio test/endtoend/wdio.local.conf.js -b chrome --suite thumbnails -b firefox
# test all in chrome
npx wdio test/endtoend/wdio.local.conf.js -b chrome
```

goto-bus-stop commented on packages/@uppy/thumbnail-generator/src/index.js:
```diff
@@ -35,20 +35,21 @@ module.exports = class ThumbnailGenerator extends Plugin {
     })
 
+    onload.then(revoke, revoke)
+
```
On resolve, revoke; on reject, also revoke :)

this is a style preference I guess, behaviour is the same as the separate URL.revokeObjectURL calls. I assume this is more confusing so will revert

arturi replied:
Aa, got it now, it’s `.then(resolve, reject)`. Tiny comment should also do it.

goto-bus-stop replied:
already reverted, it's cool :)

goto-bus-stop commented:
Will wait for #1052 :) CI ignores the integration test results ATM because we were having a few different issues. hopefully that is the last of em

arturi commented:
But the checks passed now? 🤔

goto-bus-stop commented:
Yeah integration tests are run but the result is ignored, https://github.com/transloadit/uppy/commit/17359cd6ec75eb9104745e936026b3cb70e6c816 (didn't get to revert it yet because they still don't work :weary: )

goto-bus-stop commented:
Hell yeah, passing locally but almost all fail on CI :joy:

arturi commented:
🤔do you think it’s just this PR or master branch too? If e2e tests are ignored altogether, we might have broken things many times since this PR was opened, so it probably makes sense to merge this and then fix tesst in master.

goto-bus-stop commented:
same deal on master. the integration tests are now failing here for a different reason, it seems. i agree it is probably better to merge and have a failing master branch, if only to apply some more pressure towards getting this fixed :sweat_smile:

goto-bus-stop commented:
no clue what's up with edge so i'm just trying to skip the thumbnail test there for now. if ci passes like this i'll clean this branch up a bit, make sure no unintended debugging changes are in there anymore, and land

goto-bus-stop commented:
green!!!!! renamed thumbnail:ready to thumbnail:all-generated per feedback; squashed all the fix attempts into a single commit, will wait for CI one final time then merge :tada: :tada: :tada:
